lte/gateway/python/magma/pipelined/app/gw_mac_address.py

```diff
@@ -0,0 +1,124 @@
+"""
+Copyright 2022 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import ipaddress
+import logging
+from typing import Optional
+
+from lte.protos.mobilityd_pb2 import IPAddress
+from scapy.arch import get_if_addr, get_if_hwaddr
+from scapy.data import ETH_P_ALL, ETHER_BROADCAST
+from scapy.error import Scapy_Exception
+from scapy.layers.inet6 import getmacbyip6
+from scapy.layers.l2 import ARP, Dot1Q, Ether
+from scapy.sendrecv import srp1
+
+EGRESS = "egress"
+
+
+def get_gw_mac_address(ip: IPAddress, vlan: str, non_nat_arp_egress_port: Optional[str]) -> str:
+    if ip.version == IPAddress.IPV4:
+        return _get_gw_mac_address_v4(ip, vlan, non_nat_arp_egress_port)
+    if ip.version == IPAddress.IPV6:
+        if vlan == "NO_VLAN":
+            return _get_gw_mac_address_v6(ip)
+        else:
+            gw_ip = ipaddress.ip_address(ip.address)
+            logging.error("Not supported: GW IPv6: %s over vlan %d", str(gw_ip), vlan)
+            return ""
+
+
+def _get_gw_mac_address_v4(ip: IPAddress, vlan: str, non_nat_arp_egress_port: Optional[str]) -> str:
+    try:
+        gw_ip = ipaddress.ip_address(ip.address)
+        logging.debug(
+            "sending arp via egress: %s",
+            non_nat_arp_egress_port,
+        )
+        eth_mac_src = get_if_hwaddr(non_nat_arp_egress_port)
+        psrc = "0.0.0.0"
+        egress_port_ip = get_if_addr(non_nat_arp_egress_port)
+        if egress_port_ip:
+            psrc = egress_port_ip
+
+        pkt = Ether(dst=ETHER_BROADCAST, src=eth_mac_src)
+        if vlan.isdigit():
+            pkt /= Dot1Q(vlan=int(vlan))
+        pkt /= ARP(op="who-has", pdst=gw_ip, hwsrc=eth_mac_src, psrc=psrc)
+        logging.debug("ARP Req pkt %s", pkt.show(dump=True))
+
+        res = srp1(
+            pkt,
+            type=ETH_P_ALL,
+            iface=non_nat_arp_egress_port,
+            timeout=1,
+            verbose=0,
+            nofilter=1,
+            promisc=0,
+        )
+
+        if res is not None:
```

Comment:
is refactoring in scope? early return would remove one level of if-nesting

Reply:
Changed if statement for an early return.